csharp/src/Ice/InstrumentationI.cs

```diff
@@ -23,1072 +22,748 @@ public InvocationMetrics()
-                IPConnectionInfo? info = GetIPConnectionInfo();
-                if (info != null)
-                {
-                    os.Append(info.LocalAddress).Append(':').Append(info.LocalPort);
-                    os.Append(" -> ");
-                    os.Append(info.RemoteAddress).Append(':').Append(info.RemotePort);
-                }
-                else
+                try
                 {
-                    os.Append("connection-").Append(_connectionInfo);
+                    return _endpointLookups.GetObserver(new EndpointHelper(endpt),
+                        _delegate?.GetEndpointLookupObserver(endpt));
                 }
-                if (_connectionInfo.ConnectionId.Length > 0)
+                catch (Exception ex)
                 {
-                    os.Append(" [").Append(_connectionInfo.ConnectionId).Append("]");
+                    _metrics.GetLogger().Error("unexpected exception trying to obtain observer:\n" + ex);
                 }
-                _id = os.ToString();
             }
-            return _id;
+            return null;
         }
 
-        public string GetState()
+        public MetricsAdminI GetFacet() => _metrics;
+
+        public IInvocationObserver? GetInvocationObserver(
+            IObjectPrx prx,
+            string operation,
+            IReadOnlyDictionary<string, string> ctx)
         {
-            switch (_state)
+            if (_invocations.IsEnabled)
             {
-                case ConnectionState.ConnectionStateValidating:
-                    return "validating";
-                case ConnectionState.ConnectionStateActive:
-                    return "active";
-                case ConnectionState.ConnectionStateClosing:
-                    return "closing";
-                case ConnectionState.ConnectionStateClosed:
-                    return "closed";
-                default:
-                    Debug.Assert(false);
-                    return "";
+                try
+                {
+                    return _invocations.GetObserver(new InvocationHelper(prx, operation, ctx),
+                        _delegate?.GetInvocationObserver(prx, operation, ctx));
+                }
+                catch (Exception ex)
+                {
+                    _metrics.GetLogger().Error("unexpected exception trying to obtain observer:\n" + ex);
+                }
             }
+            return null;
         }
 
-        public string GetParent()
+        public void SetObserverUpdater(IObserverUpdater? updater)
         {
-            if (_connectionInfo.AdapterName != null && _connectionInfo.AdapterName.Length > 0)
+            if (updater == null)
             {
-                return _connectionInfo.AdapterName;
+                _connections.SetUpdater(null);
             }
             else
             {
-                return "Communicator";
+                _connections.SetUpdater(updater.UpdateConnectionObservers);
             }
```

Comment:
Use `_connections.SetUpdater(updater?UpdateConnectionObservers);` instead?

Reply:
I try that first but doesn't seems to work for methods